### ISSUE TITLE
Unique salt for each Bin & random delete key

### ISSUE THREAD

**Issue by rugk** [PrivateBin/PrivateBin#68]
Related to https://github.com/PrivateBin/PrivateBin/issues/20

In the security audit "2.2. Fixed Server Salt" has not been completely fulfilled. We addressed the random number generator issues and so on, but there is still one suggestion we completely ignored:

**Unique salt for each bin:**
> This would not be possible if a random
       comment salt was generated for each post, then destroyed along
       with the post when it expires. This would, however, give users
       different avatars on different posts (which may be desired).

**Better delete link generation:**
> A random delete token could be
       generated for each post, and its hash stored along with the post.
       Then, the post could only be deleted if the user can provide the
       preimage of the hash.

Thanks to https://github.com/PrivateBin/PrivateBin/pull/65 we could indeed do this as we can consider the PRNG to be secure (and possibly fast too :smile:).

Basically currently the server Salt is a very critical thing. It is used for everything and the security of not only one feature, but multiple features depends on it.
It's one Salt to rule them all... :smiley: 

As the auditor expresses it:
> Reusing the same key for two purposes is generally a bad idea.

**Comment by elrido** [comment]
I do consider these two points solved because:
- A unique encryption salt for each paste and comment, nickname, attachment is generated already since the beginning. (and they are not reused anywhere and deleted together with the paste)
- A unique salt per bin is now generated and used for the delete link.

Hence these two points are indeed solved and I close the issue as a duplicate.

**Comment by rugk** [comment]
Okay, so the delete link issue is solved, but the first part I qouted was not about the salt used for the encryption, but for VizHash (now Identicons):

Here is the full text:

> If the salt is compromised and published, anyone can reverse
       a VizHash into the corresponding IP address, even for expired
       posts (that they saved). This would not be possible if a random
       comment salt was generated for each post, then destroyed along
       with the post when it expires. This would, however, give users
       different avatars on different posts (which may be desired).

So I think the idea of the auditor was to have one avatar salt per Pad, so that different pads show different avatars.
This would prevent "tracking" users over several pads on the same server.

Currently I could see that this is not done.

**Comment by elrido** [comment]
As we had discussed, using a unique salt per bin for the vizhashs/icons would break their purpose. Just imagine if your "rugk" Identicon on Github looked different per repository. Discussions in privatebins frequently cover multiple pastes and you would like to have an indication if this is still the same person. Hence we will not have per paste salt used in Vizhash/idendicon. If an admin does not trust the icons about leaking the IP, just set icon=none. And users can simply comment without nickname and have no icon exposed.

**Comment by rugk** [comment]
> "rugk" Identicon on Github looked different per repository

Yeah... :relaxed:

> Discussions in privatebins frequently cover multiple pastes 

This certainly depends on the experience you have. And this is also the thing which differentiates it from GitHub. In my experience a Pad is created for one thing, users discuss this one thing there and that's it. But, yes, this may differ.

**Comment by elrido** [comment]
One shot topics get typically discussed by anons. When communities form, more users start to use nicknames as it then starts to matter who is who between pastes/topics. There is at least one server out there that uses zerobin 4chan stile together with a link sharing service (the original zerobin version and yes, I had asked them if they would upgrade to this fork at the time, but they were a bit on the conservative side).

**Comment by rugk** [comment]
> but they were a bit on the conservative side

Maybe we can convince them with the coming PrivateBin release. Nobody, really, nobody should use outdated software.
